feat(terminal): trigger TermRequest autocommand events

glacambre opened the pull request:
_Original PR text_:

This commit implements a new TermOSC autocommand event and has Neovim
emit this event when children of terminal buffers emit an OSC sequence
libvterm does not handle.

The TermOSC autocommand event has additional data in its v:event :
`command`, which corresponds to the OSC number (e.g. `7` in `OSC7`) and
`payload`, which is the actual content of the command.

_Added by @gpanders_ (https://github.com/neovim/neovim/pull/22159#issuecomment-1881364545):

I renamed TermOSC to TermRequest and am populating a new v: variable v:termrequest rather than v:event (this mirrors v:termresponse). v:termrequest contains the raw control sequence string instead of parsing the string into command and payload fields.

TermRequest also handles DCS requests.

I added some tests that use TermRequest to test our automatic background detection support and truecolor detection support.

glacambre committed:
feat(terminal): trigger TermOSC autocommand events

This commit implements a new TermOSC autocommand event and has Neovim
emit this event when children of terminal buffers emit an OSC sequence
libvterm does not handle.

The TermOSC autocommand event has additional data in its v:event :
`command`, which corresponds to the OSC number (e.g. `7` in `OSC7`) and
`payload`, which is the actual content of the command.

This commit is meant as a first step in the direction of closing #2252.

glacambre commented:
This is good enough to implement Vim's `autoshelldir` feature (I did that here: https://github.com/neovim/neovim/commit/146023f79d3ea39b466d69603a819193e71e824f ), however, as I'm unhappy with the way Vim's `autoshelldir` works (it changes the window's current working directory rather than the buffer's, which yields a ton of undesirable behaviors), I decided not to implement `autoshelldir` in this pull request.

I need to spend more time thinking about how to implement `autoshelldir` (I want something that behaves the same way https://github.com/neovim/neovim/pull/16740 / https://github.com/glacambre/syncwd ) but there's no reason to hold this PR hostage until then - it's already quite useful on its own.

zeertzjq commented:
Merge conflict

glacambre commented:
Thank you for letting me know, Zeertsjq :). This should be fixed now.

glacambre commented:
The MacOS failure doesn't seem to be mine.

zeertzjq commented:
Conflicting PR: #20176

sassanh commented:
I think this PR is better approaching the issue. Triggering autocommands instead of blindly passing everything to the underlying terminal emulator gives the end user more control.

There may be synchronization concerns, I think autocommands do not run immediately and there should probably be an autocommand queue which runs them **eventually** in event loop cycle. So guaranteeing the OSC output of an application reaches the underlying terminal in the right time may be hard.

For example as I discussed [here](https://github.com/neovim/neovim/pull/20176#issuecomment-1245658103), printing a hyperlink is done by sending an OSC command, printing the title of link, and then sending another OSC command to inform the terminal emulator that the hyperlink is finished. Now if these 2 OSC commands are queued and sent together after the text is already printed, it won't work as expected.

So I think to achieve good support for OSC commands we need to make sure:
1. This autocmd runs **immediately**.
2. There is a reliable way to directly and **immediately** write to terminal.

I don't know if these are already achievable or not.

P.S. Crazy idea just came to my mind and I want to know what @zeertzjq thinks about it: Maybe we can add an option to autocommands to hint Neovim whether they should run immediately or it's ok to queue them and run them **eventually**, similar concept exists in singal/slot construct of Qt: https://doc.qt.io/qt-6/qt.html#ConnectionType-enum

glacambre commented:
>There may be synchronization concerns

Indeed, this isn't something I cared about at all (my goal was to implement something good enough to support OSC7, I did not think about other OSC sequences). I agree that it should be taken into account before anything gets merged though.

The problem can be demonstrated with the two following pieces of code:
```lua
the_lines = {'\n','\n'}
vim.api.nvim_create_autocmd({ 'TermOSC' }, {
  callback = function(e)
    table.insert(the_lines,"\npayload:" .. vim.v.event.payload .. "\nbuffer:" .. table.concat(vim.api.nvim_buf_get_lines(0, 0, -1, false), ","))
    vim.g.the_lines = the_lines
  end
})
```
```sh
for i in $(seq 1 1000) ; do
    echo "$i"
    printf "\033]7;$i\033\\"
done
```
Printing `the_lines` shows that the first 500 autocommands see an empty buffer and the remaining 500 autocommands see a buffer containing the entire output of the script. This is wrong, we want each autocommand event to see an additional line.

I could be wrong, but fixing this sounds like a lot of very hard work, so I won't do it myself (for now). Thus, anyone who wants to work on this should feel free to do so.

The current state of my TermOSC autocommand is good enough for my needs though, so the branch will get periodically rebased when I want something from master. But don't take this as a sign of activity :).

justinmk commented:
Is there a more "semantic" / less obscure name we can use, maybe `TermCmd` or `TermRequest` ?

glacambre commented:
>Is there a more "semantic" / less obscure name we can use, maybe TermCmd or TermRequest ?

OSC stands for Operating System Command, so TermCmd would be quite fitting I think. `TermSequence` or `TermUnhandled` could fit too probably.

>If the event handler is a Lua function, we could treat it as :help vim.in_fast_event() ?

Is it just a matter of calling `nlua_pcall` with `in_fast_callback` > 0? If yes it sounds like it might be within my reach, I could try that :)

justinmk commented on src/nvim/terminal.c:
```diff
@@ -169,6 +170,50 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+  .dcs = NULL,
+  .apc = NULL,
+  .pm = NULL,
+  .sos = NULL,
```
If the event is named `TermCode` we could "osc", "dcs", etc in the event "pattern".

gpanders commented:
> Is there a more "semantic" / less obscure name we can use, maybe `TermCmd` or `TermRequest` 

`TermRequest` is nicely symmetric to `TermResponse` :)

gpanders commented on src/nvim/terminal.c:
```diff
@@ -169,6 +170,50 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+
+static int on_osc(int command, VTermStringFragment frag, void *user)
+{
+  char *str = xmemdupz(frag.str, frag.len);
+  multiqueue_put(main_loop.events, emit_osc_event, 3, command, str, frag.len);
```
Instead of passing the command as a separate field from the payload, I suggest putting the command back into the sequence so that TermRequest handlers can parse the OSC sequence directly (this is how TermResponse works).

So e.g. instead of checking `v:event.command` and `v:event.payload`, have only `v:event.termrequest` (or make a `v:termrequest` analogous to `v:termresponse`? Or maybe `v:termresponse` should be moved into `v:event`?) which contains the `\x1b]{command}` prefix.

```suggestion
  StringBuilder request = KV_INITIAL_VALUE;
  kv_printf(request, "\x1b]%d%s", command, frag.str);
  multiqueue_put(main_loop.events, emit_osc_event, 1, request.items, request.size);
```

acomagu commented:
Is `on_stdout` option of `termopen` not enough? This seems synchronous.

![image](https://github.com/neovim/neovim/assets/1362750/57db86dd-537d-4731-b26e-45945727d335)

```vim
func GotOutput(job_id, msg, event)
  echo a:msg
  if match(a:msg, "\<Esc>]133;") != -1
    cal chansend(a:job_id, 'abc')
  endif
endfunc
new
let shell_job = termopen(["/bin/bash"], #{
        \   on_stdout: function('GotOutput'),
        \   on_stderr: function('GotOutput'),
        \ })
```

acomagu commented:
OSC7 Version

![image](https://github.com/neovim/neovim/assets/1362750/91806b58-42d3-4982-a0e8-793a665f0868)

```vim
func GotOutput(job_id, msg, event)
  let matches = matchlist(a:msg, "\<Esc>]7;file://\\(.*\\)\<Esc>\\\\")
  if len(matches) > 1
    echo matches[1]
  endif
endfunc
new
let shell_job = termopen(["/bin/bash"], #{
        \   on_stdout: function('GotOutput'),
        \   on_stderr: function('GotOutput'),
        \ })
```

gpanders commented on src/nvim/terminal.c:
```diff
@@ -169,6 +170,50 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+
+  tv_dict_set_keys_readonly(dict);
+
+  apply_autocmds(EVENT_TERMOSC, NULL, NULL, false, curbuf);
```
Would be useful to also include the terminal buffer number or channel number in the `data` dictionary to the autocommand callback.

gpanders replied:
The buffer number is already included and it is easy to get the `'channel'` from that, so this isn't necessary.

gpanders committed:
Rename TermOSC to TermRequest

Also add v:termrequest (analogous to v:termresponse). Populate
v:termrequest with the full sequence, rather than populating v:event.
Pass the request payload to the autocommand callback.

gpanders committed:
Support DCS requests in TermRequest

gpanders committed:
Use TermRequest in tests

gpanders commented:
Ok I've pushed a few changes. Thanks @glacambre for getting this started and sorry that it's taken almost a year to get merged.

I renamed `TermOSC` to `TermRequest` and am populating a new `v:` variable `v:termrequest` rather than `v:event` (this mirrors `v:termresponse`). `v:termrequest` contains the raw control sequence string instead of parsing the string into `command` and `payload` fields.

`TermRequest` also handles DCS requests.

I added some tests that use `TermRequest` to test our automatic background detection support and truecolor detection support.

gpanders committed:
Update docs

gpanders committed:
Fix formatting

gpanders committed:
fixup

wookayin commented on src/nvim/vvars.lua:
```diff
@@ -771,12 +771,20 @@ M.vars = {
       another escape sequence.
     ]=],
   },
+  termrequest = {
```
Please consider adding a type (maybe v:termresponse could also get one).

```suggestion
  termrequest = {
    type = "string",
```

wookayin commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -2736,13 +2737,61 @@ describe('TUI', function()
       unchanged = true,
     }
   end)
+
```
Not sure where would be the best place, but can we also add a test case for `v:termrequest` (which is currently missing apparently) --- as well as `args.data`?

gpanders replied:
The new test in `terminal/buffer_spec.lua` tests `v:termrequest`. The new tests in `tui_spec.lua` test `args.data`.

zeertzjq commented on test/functional/terminal/buffer_spec.lua:
```diff
@@ -317,6 +317,32 @@ describe(':terminal buffer', function()
     )
   end)
+
+  it('emits #OSC events', function()
```
```suggestion
  it('emits TermRequest events', function()
```

gpanders committed:
Add type to v:termresponse and v:termrequest

gpanders committed:
Test v:termrequest

gpanders committed:
make doc

justinmk commented on src/nvim/terminal.c:
```diff
@@ -169,6 +169,54 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+  }
+
+  StringBuilder request = KV_INITIAL_VALUE;
+  kv_printf(request, "\x1bP%*s", (int)commandlen, command);
```
coverity warnings, do we need to assert?

```
*** CID 477623:  Null pointer dereferences  (FORWARD_NULL)
/src/nvim/terminal.c: 205 in on_dcs()
199       if ((command == NULL) || (frag.str == NULL)) {
200         return 0;
201       }
202     
203       StringBuilder request = KV_INITIAL_VALUE;
204       kv_printf(request, "\x1bP%*s", (int)commandlen, command);
>>>     CID 477623:  Null pointer dereferences  (FORWARD_NULL)
>>>     Passing null pointer "request.items + request.size" to "memcpy", which dereferences it.
205       kv_concat_len(request, frag.str, frag.len);
206       multiqueue_put(main_loop.events, emit_term_request, request.items, (void *)request.size);
207       return 1;
208     }
209     
210     static VTermStateFallbacks vterm_fallbacks = {
```

gpanders replied:
Interesting, I don't think `request` should ever be `NULL` after `kv_printf`. Surely we don't need to `assert` after every `kv_printf` call?

justinmk replied:
> Surely we don't need to `assert` after every `kv_printf` call?

i think coverity is looking at this and thinking `kv_printf` could return early: https://github.com/neovim/neovim/blob/2dc439c672facaeb8e51ef6aa20efc0e7092eee2/src/nvim/strings.c#L2144

we could add an assert in `kv_concat_len` ?

gpanders replied:
https://github.com/neovim/neovim/pull/27022

justinmk commented on src/nvim/terminal.c:
```diff
@@ -169,6 +169,54 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
+  }
+
+  StringBuilder request = KV_INITIAL_VALUE;
+  kv_printf(request, "\x1b]%d;", command);
```
````
*** CID 477624:  Null pointer dereferences  (FORWARD_NULL)
/src/nvim/terminal.c: 192 in on_osc()
186       if (frag.str == NULL) {
187         return 0;
188       }
189     
190       StringBuilder request = KV_INITIAL_VALUE;
191       kv_printf(request, "\x1b]%d;", command);
>>>     CID 477624:  Null pointer dereferences  (FORWARD_NULL)
>>>     Passing null pointer "request.items + request.size" to "memcpy", which dereferences it.
192       kv_concat_len(request, frag.str, frag.len);
193       multiqueue_put(main_loop.events, emit_term_request, request.items, (void *)request.size);
194       return 1;
195     }
196     
197     static int on_dcs(const char *command, size_t commandlen, VTermStringFragment frag, void *user)
